### ISSUE TITLE
`transport_failure_callback` Bug

### ISSUE THREAD

**Issue by aryeh-looker** [getsentry/raven-ruby#508]
tl;dr: `transport_failure_callback` is not called on 4xx or 5xx errors and should have a failure reason

---

The idea behind `transport_failure_callback` is just fantastic. It means that there is no longer a single point of failure for reporting events to Sentry. Yet there is a bug with its implementation and I feel that it is missing a key feature. 

---

### Bug

According to the documentation, `transport_failure_callback` is called "If the transport fails to send an event to Sentry for any reason (either the Sentry server has returned a 4XX or 5XX response)." Yet it would appear that the callback is only called when `Faraday#post` raises an exception (or if `ClientState#should_try?` is false); thus `failed_send` and (more importantly) `transport_failure_callback` is not called when Sentry servers return with 4xx or 5xx status codes; the source code in question:

```
      begin
        transport.send_event(generate_auth_header, encoded_data,
                       :content_type => content_type)
        successful_send
      rescue => e
        failed_send(e, event)    # this is called _only_ if an error is raised by `Faraday#post`
        return
      end
```

I double-checked this hypothesis:

* by writing an little script that uses Faraday when contacting a server that returns 200/400/500 codes; in all cases, Faraday returns a response without raising an exception
   * script: https://gist.github.com/aryeh-looker/515632a6da7661df908a4ddd77186402
   * example script run (would expect the word "raised error" instead of 400, 500 in the script's output): https://gist.github.com/aryeh-looker/277a196269e9cfc8755f8315f29f39e2
* by using Sentry under debugger control with a DSN that contacts a server that always returns a 500 code; in that case, `transport.send_event` does not raise an exception

---

### Key Feature

It would be ideal if `transport_failure_callback` were called with the raven event and the reason why the Sentry server was not reachable or contacted; for example: 400 status, server is completely unreachable, raven client is in a failure state, the secret key is missing from the DSN, etc.

**Comment by seanlinsley** [comment]
It should be possible to have unit tests for this by using something like Webmock.

**Comment by nateberkopec** [comment]
@aryeh-looker What adapter are you using for Faraday? The default is Net/HTTP.

[I don't think Faraday is supposed to work the way you describe.](https://github.com/lostisland/faraday/blob/80c0e66102168939df04682573bc267e089a43ae/lib/faraday/response/raise_error.rb#L12) I'm working on a test case.

**Comment by nateberkopec** [comment]
Aha! The file linked is an optional middleware. I'll have this fixed in a jiffy.
